crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
     }
 }
 
+/// Representation of an exclusive clone of an upstream repo
+///
+/// [UpstreamRepo] and [UpstreamCatalog] ensure safe access to individual catalog branches.
+/// Every [UpstreamRepo] instance represents an exclusive clone
+/// and can only ever create a single [UpstreamCatalog] instance at a time.
+struct UpstreamRepo(Git);
+
+impl UpstreamRepo {
+    /// Clone the upstream repo
+    async fn clone_repo(
+        publish_ref: &PublishRef,
+        temp_dir: impl AsRef<Path>,
+    ) -> Result<Self, PublishError> {
+        let url = publish_ref.clone_url();
+        let repo_dir = tempfile::tempdir_in(temp_dir).unwrap().into_path(); // todo catch error
+        let repo = <Git as GitProvider>::clone(&url, &repo_dir, false).await?;
+
+        Ok(Self(repo))
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
+    }
+
+    /// Create an [UpstreamCatalog] by checking out or creating a catalog branch.
+    ///
+    /// `Git` objects can switch branches at any time leaving the repo in an unknown state.
+    /// [get_catalog] ensures that only one [UpstreamCatalog] exists at a time by requiring a `&mut self`.
+    async fn get_catalog(&mut self, system: &System) -> Result<UpstreamCatalog, PublishError> {
+        if self.0.list_branches().await? // todo: catch error
+            .into_iter().any(|info| info.name == Self::catalog_branch_name(system))
+        {
+            self.0
+                .checkout(&Self::catalog_branch_name(system), false)
+                .await?; // todo: catch error
+        } else {
+            self.0
+                .checkout(&Self::catalog_branch_name(system), true)
+                .await?;
+            self.0
+                .set_origin(&Self::catalog_branch_name(system), "origin")
+                .await?;
+        }
+        Ok(UpstreamCatalog(&self.0))
+    }
+}
+
+/// Representation of a specific catalog branch in an exclusive clone of an upstream repo.
+///
+/// [UpstreamCatalog] guaranteesd that during its lifetime all operations on the underlying git repo
+/// are performed on a single branch,
+/// and that the branch is pushed to upstream before a new branch can be checked out.
```

Comment:
Why do you think its undesirable?
I can only think of publishing multiple systems at once as a use case to checkout other branches but flox is squarely running with a single system context, and mutliple systesm was not a requirement of publish so far.


If we want this it should be easy to implement it by running an `--atomic` push of all (or all modified) catalog branches on the `UpstreamRepo` type and return back control with `add_snapshot` or a separate method that destructs `UpstreamCatalog`

Reply:
Depending on how we setup builders, I would expect a possibly quite common use case for publish would be to run `flox publish --systems x86_64-linux aarch64-darwin` in CI. I agree that's not a requirement at the moment, but it seems like creating abstractions to explicitly prevent something we'll likely want to do in the future is a waste of effort. That said maybe it will make things safer for now. Doesn't have to be a blocker.

Reply:
Doing multiple system publishes would change the flow anyway.
Changing when to drop `CatalogChannel` is a trivial change compared to that.